api/views/__init__.py

```diff
@@ -153,16 +173,36 @@ class DomainAddressViewSet(SaveToRequestUser, viewsets.ModelViewSet):
     def get_queryset(self):
         return DomainAddress.objects.filter(user=self.request.user)
 
-    def perform_create(self, serializer):
+    def perform_create(self, serializer: BaseSerializer[DomainAddress]) -> None:
         try:
             serializer.save(user=self.request.user)
         except IntegrityError:
+            assert isinstance(self.request.user, User)
             domain_address = DomainAddress.objects.filter(
                 user=self.request.user, address=serializer.validated_data.get("address")
             ).first()
-            raise ConflictError(
-                {"id": domain_address.id, "full_address": domain_address.full_address}
-            )
+            existing_id = existing_address = None
+            if domain_address:
+                existing_id = domain_address.id
+                existing_address = domain_address.full_address
+            raise DomainAddressConflictError(existing_id, existing_address)
+
+        if not isinstance(serializer, DomainAddressSerializer) or not isinstance(
+            self.request.user, User
+        ):
+            return
+        if fxa := self.request.user.profile.fxa:
+            mozilla_accounts_id = fxa.uid
+        else:
+            mozilla_accounts_id = ""
+        glean_logger().record_email_generate_mask(
+            user_agent=self.request.headers.get("user-agent", ""),
+            ip_address=self.request.headers.get("ip-address", ""),
+            mozilla_accounts_id=mozilla_accounts_id,
+            has_generated_for=bool(serializer.data.get("generated_for", None)),
+            is_random_mask=False,
+            created_by_api=True,
```

Comment:
same question (non-blocking): in API `perform_create` or in a `post_save` signal handler?

Reply:
The `post_save` signal doesn't have access to the request headers